### PR TITLE
Normalize naming of min/max arguments across widgets

### LINKED ISSUE #1999: Min/max API consistency
The following widgets have a concept of min/max:
* Slider
* NumberInput
* DateInput
* TimeInput

Affter #1951 is merged, all of them will have `min_value` and `max_value` properties, except for Slider, which has a combined `range` property. For simplicity, I suggest changing all of these to simply `min` and `max`, and likewise with the constructor arguments.

This should be done before the next Toga release, to avoid an additional layer of backward compatibility code for DateInput and TimeInput on top of the existing `min_date`, `max_date`, etc.

Also, as mentioned at https://github.com/beeware/toga/pull/1951#issuecomment-1597789955, NumberInput should be changed so that setting an inconsistent min/max will cause one to clip the other, rather than a ValueError.

### PR DESCRIPTION
**Pull request by freakboy3742**
* Slider now accepts `min`/`max`, rather than `range`.
* NumberInput now uses `min`/`max`, rather than `min_value`/`max_value`
* NumberInput now clips inconsistent `min`/`max` values
* DateInput now uses `min`/`max`, rather than `DatePicker.min_date`/`max_date`
* TimeInput now uses `min`/`max`, rather than `TimePicker.min_time`/`max_time`
* Restores 100% core test coverage for DateInput and TimeInput (caused by a missing code path in the deprecation shim)
* Restores 100% cocoa test coverage for Button (caused by the recent change to interpret default widget height as NONE, rather than 0)
* Ports the toga.platform tests to pytest. Not sure why, but the introduction of the NotImplmented factory shim in #1992 caused local pytest runs to fail for me. Tox runs were fine; but since we needed to update the tests anyway, now was as good a time as any. This also corrects a minor issue with the definition of `current_platform`, and gets the module to 100% test coverage.

Backwards compatibility shims have been retained for old arguments, accessors and mutators.

Fixes #1999.

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: testbed/tests/widgets/test_textinput.py [hunk 1]
```diff
@@ -94,7 +94,7 @@ async def test_on_change_user(widget, probe, on_change):
 
     for count, char in enumerate("Hello world", start=1):
         await probe.type_character(char)
-        await probe.redraw(f"Typed {char!r}")
+        await probe.redraw(f"Typed {char!r}", delay=0.02)
```
Comment (mhsmith) [inline testbed/tests/widgets/test_textinput.py:97]: This should have a comment explaining what the delay is for, and what would happen without it.

File: testbed/tests/widgets/test_slider.py [hunk 2]
```diff
@@ -103,16 +106,17 @@ async def test_change(widget, probe, on_change):
 async def test_min(widget, probe, on_change):
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply testbed/tests/widgets/test_slider.py:106]: And likewise with the comment above `test_max`.

File: core/tests/widgets/test_dateinput.py [hunk 3]
```diff
@@ -223,42 +222,65 @@ def test_invalid_max_value(widget, value, exc, message):
...
+        widget.min_date = MIN
+        assert widget.min_date == MIN
+    assert widget.min == MIN
+
+    with pytest.warns(
+        DeprecationWarning, match="DatePicker.max_date has been renamed DateInput.max"
+    ):
+        widget.max_date = MAX
+        assert widget.max_date == MAX
+    assert widget.max == MAX
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply core/tests/widgets/test_dateinput.py:271]: The rest of the test isn't redundant - this is the edge case that was missing coverage in the original test suite. The issue is constructing Date/TimePicker (not DateTime/Input) but *without* min/max initial values; the `KeyError: pass` edge case wasn't tested. Granted, we don't need to use `min_date` and `max_date` as the properties to test this worked afterwards, but we have to us *a* property, so we might as well use the deprecated ones.

File: testbed/tests/widgets/test_slider.py [hunk 4]
```diff
@@ -78,13 +80,14 @@ async def test_value(widget, probe, on_change):
 def assert_set_value(widget, value_in, value_out=None):
     if value_out is None:
         value_out = value_in
-    value_out = assert_set_get(widget, "value", value_in, value_out)
+    value_out = assert_set_get(widget, "value", value_in, pytest.approx(value_out))
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply testbed/tests/widgets/test_slider.py:83]: A mistake on my part. In porting from range to min/max, I missed that the iOS implementation was maintaining a shadow copy of the range because of 32 bit precision.

File: testbed/tests/widgets/test_slider.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (freakboy3742) [reply testbed/tests/widgets/test_slider.py:106]: I think these are still accurate - it's referring to the bulk of the range tests being in toga.core, not the range tests in testbed.

[the remark this answers is not in the document]
Comment (mhsmith) [reply testbed/tests/widgets/test_slider.py:106]: OK, I'll update it to clarify that.

**Review by mhsmith — APPROVED** [review]
